### PR TITLE
feat: support different types for `CompatReader`

### PR DESCRIPTION
**Pull request by KKould**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/issues/3517

## What's changed and what's your intention?

Support CompatReader to be compatible with different types of reads (for modify column type)

Whether the type allows conversion will be verified in the next pr: (reference: 
`ModifyColumn::validate` https://github.com/GreptimeTeam/greptimedb/pull/3701/files#diff-839e23ed322c90eacc42fdddeb1f7108335e6434eba0a7a1c6dbc42052a8b9e9)

Tips: Please closely [64FD8R166](https://github.com/GreptimeTeam/greptimedb/pull/3745/files#diff-f6ecbd6cf3bbc3d92c1993aeed2c4af70186b8014af7a3a14798ac52a1664fd8R166) in the `cast` in this line of code, the `safe` of the `CastOptions` it used by default is true, so it is guaranteed that the type is allowed to be transformed and specific When the value cannot be converted, set the value to null, e.g. type: `String` -> `Integer`, value: `"1"` -> `1`, `"hello"` -> `null`

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/mito2/src/read/compat.rs [hunk 1]
```diff
@@ -85,7 +86,7 @@ pub(crate) fn has_same_columns(left: &RegionMetadata, right: &RegionMetadata) ->
     }
 
     for (left_col, right_col) in left.column_metadatas.iter().zip(&right.column_metadatas) {
-        if left_col.column_id != right_col.column_id {
+        if left_col != right_col {
```
Comment (waynexia) [inline src/mito2/src/read/compat.rs:89]: Only comparing what you need can be a little faster. And the `ColumnMetadata` contains much information we don't care about here like name, metadata, semantic type etc. Can you make a util method to only compare whether we need cast types on read?

**Comment by KKould** [comment]
> We can also set up a simple rule.
> - fill NULL if the column is nullable
> - fill default if the column is not null

https://github.com/GreptimeTeam/greptimedb/issues/3517#issuecomment-2063190801

`cast` can only be converted to null but cannot set default. Do you have any suggestions on this?

File: src/mito2/src/read/compat.rs [hunk 2]
```diff
@@ -239,9 +254,10 @@ fn may_compat_fields(
     mapper: &ProjectionMapper,
     actual: &RegionMetadata,
 ) -> Result<Option<CompatFields>> {
+    let expect = mapper.metadata();
     let expect_fields = mapper.batch_fields();
     let actual_fields = Batch::projected_fields(actual, mapper.column_ids());
-    if expect_fields == actual_fields {
+    if expect.schema == actual.schema {
```
Comment (evenyag) [inline src/mito2/src/read/compat.rs:260]: If we only read a subset of fields (projection) then this check will also return false. We can returns the column id and data type in `batch_fields()` and `projected_fields()` above.

File: src/mito2/src/read/compat.rs [hunk 3]
```diff
@@ -156,7 +157,21 @@ impl CompatFields {
...
             .iter()
             .map(|index_or_default| match index_or_default {
-                IndexOrDefault::Index(index) => batch.fields()[*index].clone(),
+                IndexOrDefault::Index { pos, cast_type } => {
+                    let old_column = &batch.fields()[*pos];
+
+                    let data = if let Some(ty) = cast_type {
+                        // Safety: We ensure type can be converted and the new batch should be valid.
+                        // Tips: `safe` must be true in `CastOptions`, which will replace the specific value with null when it cannot be converted.
+                        old_column.data.cast(ty).unwrap()
```
Comment (evenyag) [inline src/mito2/src/read/compat.rs:166]: We need test cases to ensure updating arrow won't break this. We could test some behavior in `VectorOp::cast()`'s test cases.

File: src/mito2/src/read/compat.rs [hunk 4]
```diff
@@ -255,8 +271,21 @@ fn may_compat_fields(
         .iter()
         .map(|column_id| {
             if let Some(index) = source_field_index.get(column_id) {
+                // TODO(kould): more Modify case, e.g. column name
```
Comment (evenyag) [inline src/mito2/src/read/compat.rs:274]: We use column id internally so modifying the column name should not affect this.

File: src/mito2/src/read/compat.rs [hunk 5]
```diff
@@ -255,8 +271,21 @@ fn may_compat_fields(
...
+                let mut modify_type = None;
+
+                // Safety: check on [`TableMeta::modify_columns`].
+                let expect_column = expect.column_by_id(*column_id).unwrap();
+                let actual_column = actual.column_by_id(*column_id).unwrap();
+
+                if expect_column != actual_column {
+                    modify_type = Some(expect_column.column_schema.data_type.clone())
+                }
```
Comment (evenyag) [inline src/mito2/src/read/compat.rs:283]: Compare the whole metadata here may introduce the same issue mentioned in https://github.com/GreptimeTeam/greptimedb/pull/3745#discussion_r1572049239

We can reuse the column type if `batch_fields()` and `projected_fields()` can return it. Then we don't need to call `column_by_id` to acquire the metadata again.

**Comment by evenyag** [comment]
> `cast` can only be converted to null but cannot set default. Do you have any suggestions on this?

Maybe we could only allow modifying the data type when the column is nullable or casting the type won't be fallible.

**Review by evenyag — COMMENTED** [review]
Sorry for the slow review. Don't have enough time to review this last week.

**Comment by KKould** [comment]
> > `cast` can only be converted to null but cannot set default. Do you have any suggestions on this?
> 
> Maybe we could only allow modifying the data type when the column is nullable or casting the type won't be fallible.

I added this check in pr: https://github.com/GreptimeTeam/greptimedb/pull/3757 https://github.com/GreptimeTeam/greptimedb/pull/3757/commits/9003afc9876665dfb71f2eb1e482b4cda7579480

File: src/mito2/src/read/compat.rs [hunk 6]
```diff
@@ -248,15 +263,23 @@ fn may_compat_fields(
...
     let index_or_defaults = expect_fields
         .iter()
-        .map(|column_id| {
-            if let Some(index) = source_field_index.get(column_id) {
+        .map(|(column_id, expect_data_type)| {
+            if let Some((index, actual_data_type)) = source_field_index.get(column_id) {
+                let mut modify_type = None;
+
+                if expect_data_type != *actual_data_type {
+                    modify_type = Some(expect_data_type.clone())
```
Comment (evenyag) [inline src/mito2/src/read/compat.rs:276]:
```suggestion
                let mut cast_type = None;

                if expect_data_type != *actual_data_type {
                    cast_type = Some(expect_data_type.clone())
```

File: src/mito2/src/read/compat.rs [hunk 7]
```diff
@@ -248,15 +263,23 @@ fn may_compat_fields(
...
+                }
                 // Source has this field.
-                Ok(IndexOrDefault::Index(*index))
+                Ok(IndexOrDefault::Index {
+                    pos: *index,
+                    cast_type: modify_type,
```
Comment (evenyag) [inline src/mito2/src/read/compat.rs:281]:
```suggestion
                    cast_type,
```

**Review by evenyag — COMMENTED** [review]
Mostly LGTM

**Review by evenyag — APPROVED** [review]
LGTM

**Review by v0y4g3r — APPROVED** [review]
LGTM!

**Pull request merged**